dex/src/main/java/io/crate/data/join/CrossJoinBlockNLBatchIterator.java

```diff
@@ -0,0 +1,167 @@
+import io.crate.data.BatchIterator;
+import io.crate.data.Row;
+import io.crate.data.UnsafeArrayRow;
+
+import java.util.ArrayList;
+
+/**
+ * This BatchIterator is used for both CrossJoins and InnerJoins as for the InnerJoins
+ * the joinCondition is tested later on as a filter Projection.
+ *
+ * The Block Nested Loop algorithm is used to implement the cross join.
+ *
+ * <pre>
+ *     // Nested Loop
+ *     for (leftRow in left) {
+ *         for (rightRow in right) {
+ *             match?
+ *         }
+ *     }
+ * </pre>
+ * <pre>
+ *     // Block Nested Loop
+ *     fill buffer with next items from the left
+ *     for (row in buffer) {
+ *         for (rightRow in right) {
+ *             match?
+ *         }
+ *     }
+ *     repeat
+ * </pre>
+ */
+public class CrossJoinBlockNLBatchIterator<L, R, C> extends JoinBatchIterator<L, R, C> {
+
+    private final BlockSizeCalculator blockSizeCalculator;
+    private final ArrayList<Object[]> blockBuffer;
+    private final UnsafeArrayRow rowWrapper;
+    private final RowAccounting rowAccounting;
+
+    private int blockBufferMaxSize;
+    private int bufferPos;
+    private boolean rightInitialized;
+
+    CrossJoinBlockNLBatchIterator(BatchIterator<L> left,
+                                  BatchIterator<R> right,
+                                  ElementCombiner<L, R, C> combiner,
+                                  BlockSizeCalculator blockSizeCalculator,
+                                  RowAccounting rowAccounting) {
+        super(left, right, combiner);
+        this.blockSizeCalculator = blockSizeCalculator;
+        this.blockBuffer = new ArrayList<>(0);
+        this.rowWrapper = new UnsafeArrayRow();
+        this.rowAccounting = rowAccounting;
+        resizeBlockBuffer();
+    }
+
+    private void resizeBlockBuffer() {
+        rowAccounting.release();
+        blockBufferMaxSize = blockSizeCalculator.calculateBlockSize();
+        blockBuffer.clear();
+        blockBuffer.ensureCapacity(blockBufferMaxSize);
+        bufferPos = -1;
+    }
+
+    @Override
+    public void moveToStart() {
+        left.moveToStart();
+        right.moveToStart();
+        activeIt = right;
+        rightInitialized = false;
+        resizeBlockBuffer();
+    }
+
+    @Override
+    public boolean moveNext() {
+        do {
+            if (bufferPos == -1) {
+                // block buffer needs to be filled
+                activeIt = right;
+                // try advancing the right side first to check if we have items on the right
+                if (!rightInitialized) {
+                    if (right.moveNext()) {
+                        rightInitialized = true;
+                    } else {
+                        return false;
+                    }
+                }
+                activeIt = left;
+                while (blockBuffer.size() < blockBufferMaxSize) {
+                    if (left.moveNext()) {
+                        Row row = (Row) left.currentElement();
```

Comment:
I don't think it's nice to pretend to support generic items and then having a hard dependency on the items being castable to `Row` at runtime.
I think we've to change the `RowAccounting` class to support generics too or make the implementation completely concrete and no longer generic. Same for the blockBuffer logic.

Another option would be to parameterize the whole `blockBuffer` logic and let the use-site use a "row accounting" `blockBuffer.`

Reply:
I've simply replaced the generic arguments of `CrossJoinBlockNLBatchIterator` with `Row`. So the cast is not necessary anymore.